File: packages/jetbrains-plugin/src/main/java/com/chainlesschain/ide/CapabilityNegotiation.java
package com.chainlesschain.ide;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Java twin of packages/cli/src/lib/capability-negotiation.js — bidirectional
 * capability negotiation with N / N-1 downgrade for Agent Protocol
 * (agent-sdk docs/PROTOCOL.md §1.2.2 / §1.3).
 *
 * <p>The JetBrains panel is a CLIENT: it builds a {@code hello} offer and reads
 * the CLI's {@code system/negotiated} reply. Keeping the SAME algorithm here
 * lets either side reason about the agreed level, and the shared fixture
 * ({@code packages/cli/__tests__/fixtures/capability-negotiation-cases.json},
 * asserted by {@code CapabilityNegotiationSmokeMain} + the JS
 * {@code capability-negotiation.test.js}) keeps the two honest.
 *
 * <p>Pure logic — no IntelliJ SDK. Mirrors the JS rules exactly:
 * agreedVersion = min(maxes); incompatible when below max(mins); effective
 * features = server-offered ∩ client-accepted minus version-gated; no client
 * offer keeps full behavior.
 */
public final class CapabilityNegotiation {

    public static final int PROTOCOL_VERSION = 1;
    public static final int PROTOCOL_MIN_VERSION = 1;
    public static final List<String> PROTOCOL_FEATURES =
            Collections.unmodifiableList(
                    Arrays.asList("event_seq", "tool_use_id", "trace_id"));

    private CapabilityNegotiation() {}

    /** Negotiation outcome (fields mirror the JS result object). */
    public static final class Result {
        public final boolean ok;
        /** Agreed version, or null when the ranges do not overlap. */
        public final Integer agreedVersion;
        public final List<String> features;          // sorted
        public final boolean downgraded;
        public final List<String> disabledFeatures;  // sorted
        public final boolean clientAware;
        public final String reason;

        Result(boolean ok, Integer agreedVersion, List<String> features,
               boolean downgraded, List<String> disabledFeatures,
               boolean clientAware, String reason) {
            this.ok = ok;
            this.agreedVersion = agreedVersion;
            this.features = features;
            this.downgraded = downgraded;
            this.disabledFeatures = disabledFeatures;
            this.clientAware = clientAware;
            this.reason = reason;
        }
    }

    private static int intOr(Object v, int fallback) {
        if (v instanceof Number) {
            double d = ((Number) v).doubleValue();
            if (d == Math.floor(d) && d > 0 && !Double.isInfinite(d)) {
                return (int) d;
            }
        }
        return fallback;
    }

    private static boolean truthy(Object v) {
        if (v == null) return false;
        if (v instanceof Boolean) return (Boolean) v;
        if (v instanceof Number) return ((Number) v).doubleValue() != 0;
        if (v instanceof String) return !((String) v).isEmpty();
        return true;
    }

    /** Array of strings, or truthy keys of an object → sorted de-duplicated list. */
    static List<String> normalizeFeatureList(Object input) {
        LinkedHashSet<String> keys = new LinkedHashSet<>();
        if (input instanceof List) {
            for (Object o : (List<?>) input) {
                if (o instanceof String && !((String) o).isEmpty()) keys.add((String) o);
            }
        } else if (input instanceof Map) {
            for (Map.Entry<?, ?> e : ((Map<?, ?>) input).entrySet()) {
                if (truthy(e.getValue())) keys.add(String.valueOf(e.getKey()));
            }
        } else {
            return new ArrayList<>();
        }
        List<String> out = new ArrayList<>(keys);
        Collections.sort(out);
        return out;
    }

    private static boolean versionOk(String feature, int version,
                                     Map<String, Integer> mins) {
        Integer m = mins == null ? null : mins.get(feature);
        return (m == null ? 0 : m) <= version;
    }

    /**
     * @param server         {@code {protocolVersion, minProtocolVersion?, features}}
     * @param clientOffer     null (legacy peer) or {@code {protocolVersion?, minProtocolVersion?, features?}}
     * @param featureMinVersion optional feature → minimum version gate (may be null)
     */
    public static Result negotiate(Map<String, Object> server,
                                   Map<String, Object> clientOffer,
                                   Map<String, Integer> featureMinVersion) {
        int serverMax = intOr(server.get("protocolVersion"), PROTOCOL_VERSION);
        int serverMin = intOr(server.get("minProtocolVersion"), serverMax);
        List<String> serverFeatures = normalizeFeatureList(server.get("features"));

        if (clientOffer == null) {
            List<String> feats = new ArrayList<>();
            for (String f : serverFeatures) {
                if (versionOk(f, serverMax, featureMinVersion)) feats.add(f);
            }
            return new Result(true, serverMax, feats, false,
                    new ArrayList<>(), false, null);
        }

        int clientMax = intOr(clientOffer.get("protocolVersion"), serverMax);
        int clientMin = intOr(clientOffer.get("minProtocolVersion"), clientMax);
        int agreedVersion = Math.min(serverMax, clientMax);
        int floor = Math.max(serverMin, clientMin);

        if (agreedVersion < floor) {
            List<String> disabled = new ArrayList<>(serverFeatures);
            Collections.sort(disabled);
            return new Result(false, null, new ArrayList<>(), true, disabled, true,
                    "no common protocol version (server " + serverMin + "-" + serverMax
                            + ", client " + clientMin + "-" + clientMax + ")");
        }

        // A client that omits `features` accepts whatever the agreed version
        // offers; present (even empty) narrows to the intersection.
        Set<String> clientFeatures = clientOffer.containsKey("features")
                ? new HashSet<>(normalizeFeatureList(clientOffer.get("features")))
                : null;

        List<String> enabled = new ArrayList<>();
        List<String> disabled = new ArrayList<>();
        for (String f : serverFeatures) {
            boolean okVersion = versionOk(f, agreedVersion, featureMinVersion);
            boolean okClient = clientFeatures == null || clientFeatures.contains(f);
            if (okVersion && okClient) enabled.add(f);
            else disabled.add(f);
        }
        Collections.sort(enabled);
        Collections.sort(disabled);
        boolean downgraded = agreedVersion < serverMax || !disabled.isEmpty();
        return new Result(true, agreedVersion, enabled, downgraded, disabled, true, null);
    }
}

File: packages/cli/src/lib/capability-negotiation.js
/**
 * Bidirectional capability negotiation with N / N-1 downgrade for Agent
 * Protocol (agent-sdk docs/PROTOCOL.md §1.3).
 *
 * The CLI advertises what it can speak one-directionally (buildAgentCapabilities
 * → `cc agent --capabilities`, and the `system/init` line's protocol_version).
 * That is not negotiation: a client (VS Code / JetBrains panel) had no way to
 * announce what IT understands, and the CLI had no rule to pick a common level
 * or step DOWN when the two disagree. When the protocol later grows a v2 line
 * shape, a v1-only client would silently mis-parse it.
 *
 * This module is the missing algorithm. A client MAY send a `hello` as its
 * first stream-json input line:
 *   {"type":"hello","protocol_version":2,"min_protocol_version":1,
 *    "features":["event_seq","trace_id"]}
 * The CLI negotiates it against its own offer and echoes the agreed level as a
 * `system/negotiated` line, then honors it for the rest of the run.
 *
 * Rules:
 *   - agreedVersion = min(server.max, client.max)  — never above what either
 *     can parse.
 *   - Incompatible when agreedVersion < max(server.min, client.min): the two
 *     version ranges don't overlap. ok:false — the caller keeps a safe baseline
 *     or errors out; it never speaks a version the peer can't read.
 *   - effective features = the server-offered features the client also accepts
 *     (intersection), minus any whose minimum protocol version is above the
 *     agreed version (an N-only field is dropped for an N-1 session).
 *   - No client offer at all → legacy client that predates negotiation. The
 *     additive fields are defined "consumers MUST tolerate absence AND
 *     presence" (docs/PROTOCOL.md §1.2.1), so we keep FULL behavior unchanged
 *     (byte-for-byte) and mark clientAware:false. Negotiation only ever
 *     RESTRICTS when a client explicitly narrows the set.
 *
 * Pure: no fs / process / clock. The same algorithm is mirrored in the Java
 * twin (jetbrains-plugin CapabilityNegotiation.java); both read the shared
 * fixture __tests__/fixtures/capability-negotiation-cases.json.
 */

/** The current protocol version the CLI speaks (mirror of PROTOCOL_VERSION). */
export const PROTOCOL_VERSION = 1;

/**
 * The oldest protocol version the CLI can still speak end-to-end (the N-1 in
 * "N / N-1"). At v1 there is no older line shape, so min === current; once v2
 * ships this drops to 1 so a v1-only client negotiates a v1 session.
 */
export const PROTOCOL_MIN_VERSION = 1;

/**
 * The wire-protocol features subject to negotiation — the additive per-line
 * fields a client may or may not understand. Runtime capabilities (bare,
 * worktree, mcp, …) are NOT negotiated: they change what the CLI can do, not
 * the shape of a line the client must parse.
 */
export const PROTOCOL_FEATURES = ["event_seq", "tool_use_id", "trace_id"];

/**
 * Minimum protocol version a feature requires. Empty today (every current
 * feature is a v1 additive field). When a v2 field lands, add it here so an
 * N-1 (v1) session drops it automatically.
 * @type {Record<string, number>}
 */
export const FEATURE_MIN_VERSION = {};

function intOr(value, fallback) {
  const n = Number(value);
  return Number.isInteger(n) && n > 0 ? n : fallback;
}

/**
 * Coerce a features declaration into a sorted, de-duplicated array of keys.
 * Accepts an array of strings, or an object whose truthy keys are the features
 * (the nested manifest shape). Anything else → [].
 * @returns {string[]}
 */
export function normalizeFeatureList(input) {
  let keys;
  if (Array.isArray(input)) {
    keys = input.filter((k) => typeof k === "string" && k);
  } else if (input && typeof input === "object") {
    keys = Object.keys(input).filter((k) => input[k]);
  } else {
    return [];
  }
  return [...new Set(keys)].sort();
}

/**
 * Extract the server's negotiation offer from a `cc agent --capabilities`
 * manifest (buildAgentCapabilities output). Only the negotiable wire features
 * the manifest advertises truthy are offered.
 * @param {object} manifest
 * @returns {{protocolVersion:number, minProtocolVersion:number, features:string[]}}
 */
export function buildServerOffer(manifest = {}) {
  const protocolVersion = intOr(manifest.protocol_version, PROTOCOL_VERSION);
  const minProtocolVersion = intOr(
    manifest.min_protocol_version,
    Math.min(PROTOCOL_MIN_VERSION, protocolVersion),
  );
  const advertised = normalizeFeatureList(manifest.features);
  const features = PROTOCOL_FEATURES.filter((f) => advertised.includes(f));
  return { protocolVersion, minProtocolVersion, features };
}

/**
 * Negotiate the effective protocol version + feature set between a server offer
 * and a client offer.
 *
 * @param {{protocolVersion:number, minProtocolVersion?:number, features:string[]}} server
 * @param {null|{protocolVersion?:number, minProtocolVersion?:number, features?:string[]}} clientOffer
 * @param {object} [opts]
 * @param {Record<string,number>} [opts.featureMinVersion=FEATURE_MIN_VERSION]
 * @returns {{
 *   ok:boolean, agreedVersion:(number|null), features:string[],
 *   downgraded:boolean, disabledFeatures:string[], clientAware:boolean,
 *   reason:(string|null)
 * }}
 */
export function negotiateProtocol(server = {}, clientOffer = null, opts = {}) {
  const featureMinVersion = opts.featureMinVersion || FEATURE_MIN_VERSION;
  const serverMax = intOr(server.protocolVersion, PROTOCOL_VERSION);
  const serverMin = intOr(server.minProtocolVersion, serverMax);
  const serverFeatures = normalizeFeatureList(server.features);

  const versionOk = (f, v) => (featureMinVersion[f] || 0) <= v;

  // No client offer → legacy peer. Keep full behavior (only version-gate the
  // server's own features against its own max), byte-for-byte unchanged.
  if (clientOffer == null) {
    return {
      ok: true,
      agreedVersion: serverMax,
      features: serverFeatures.filter((f) => versionOk(f, serverMax)),
      downgraded: false,
      disabledFeatures: [],
      clientAware: false,
      reason: null,
    };
  }

  const clientMax = intOr(clientOffer.protocolVersion, serverMax);
  const clientMin = intOr(clientOffer.minProtocolVersion, clientMax);
  const agreedVersion = Math.min(serverMax, clientMax);
  const floor = Math.max(serverMin, clientMin);

  if (agreedVersion < floor) {
    return {
      ok: false,
      agreedVersion: null,
      features: [],
      downgraded: true,
      disabledFeatures: [...serverFeatures].sort(),
      clientAware: true,
      reason:
        `no common protocol version (server ${serverMin}-${serverMax}, ` +
        `client ${clientMin}-${clientMax})`,
    };
  }

  // A client that omits `features` accepts whatever the agreed version offers;
  // a client that sends the array narrows to the intersection.
  const clientFeatures =
    clientOffer.features === undefined
      ? null
      : new Set(normalizeFeatureList(clientOffer.features));

  const enabled = [];
  const disabled = [];
  for (const f of serverFeatures) {
    const okVersion = versionOk(f, agreedVersion);
    const okClient = clientFeatures == null || clientFeatures.has(f);
    if (okVersion && okClient) enabled.push(f);
    else disabled.push(f);
  }

  return {
    ok: true,
    agreedVersion,
    features: enabled.sort(),
    downgraded: agreedVersion < serverMax || disabled.length > 0,
    disabledFeatures: disabled.sort(),
    clientAware: true,
    reason: null,
  };
}

/** Feature key → the stream line field it gates. */
const FEATURE_TO_FIELD = {
  event_seq: "seq",
  trace_id: "trace_id",
  tool_use_id: "tool_use_id",
};

/**
 * Fold a negotiation result into a live field-gate the emitter reads per line:
 * a field stays stamped only if its feature survived negotiation. On an
 * incompatible (ok:false) result nothing is changed (the caller keeps its safe
 * baseline). Mutates `gate` in place and returns it.
 *
 * @param {{features:string[], ok:boolean}} result
 * @param {{seq?:boolean, trace_id?:boolean, tool_use_id?:boolean}} gate
 */
export function applyNegotiationToGate(result, gate = {}) {
  if (!result || result.ok === false) return gate;
  const enabled = new Set(result.features || []);
  for (const feature of PROTOCOL_FEATURES) {
    const field = FEATURE_TO_FIELD[feature];
    if (field) gate[field] = enabled.has(feature);
  }
  return gate;
}
